File: impl/js/src/subject/subjectFieldMapperCatalog.ts
import type { SubjectFieldMapper } from "./subjectFieldMapper.ts"

/**
 * A {@link SubjectFieldMapper} registered per subject name - the
 * catalog-level counterpart to attaching one directly via `createSubject`.
 * Handed to `Policy`/`PolicyBuilder` via `KeycardConfig.mapper` so field
 * mappers can be registered centrally, for subjects created without one
 * inline.
 */
export class SubjectFieldMapperCatalog {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private readonly mappers = new Map<string, SubjectFieldMapper<any>>()

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  constructor(entries: Record<string, SubjectFieldMapper<any>> = {}) {
    for (const [subjectName, mapper] of Object.entries(entries)) {
      this.register(subjectName, mapper)
    }
  }

  register<TData>(subjectName: string, mapper: SubjectFieldMapper<TData>): this {
    this.mappers.set(subjectName, mapper)
    return this
  }

  get(subjectName: string): SubjectFieldMapper<unknown> | undefined {
    return this.mappers.get(subjectName)
  }
}

File: impl/java/src/main/java/com/cptnfizzbin/keycard/subject/SubjectFieldMapperCatalog.java
package com.cptnfizzbin.keycard.subject;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;

/**
 * A {@link SubjectFieldMapper} registered per subject name - the
 * catalog-level counterpart to attaching one directly via
 * {@link SubjectFactory#create(String, SubjectFieldMapper)}. Handed to
 * {@code Policy}/{@code PolicyBuilder} via {@code KeycardConfig#getMapper()}
 * so field mappers can be registered centrally, for subjects created
 * without one inline.
 */
public final class SubjectFieldMapperCatalog {
    private final Map<String, SubjectFieldMapper<?>> mappers;

    private SubjectFieldMapperCatalog(Map<String, SubjectFieldMapper<?>> mappers) {
        this.mappers = new LinkedHashMap<>(mappers);
    }

    public static Builder builder() {
        return new Builder();
    }

    public Optional<SubjectFieldMapper<?>> get(String subjectName) {
        return Optional.ofNullable(mappers.get(subjectName));
    }

    public static final class Builder {
        private final Map<String, SubjectFieldMapper<?>> mappers = new LinkedHashMap<>();

        private Builder() {}

        public <T> Builder register(String subjectName, SubjectFieldMapper<T> mapper) {
            mappers.put(subjectName, mapper);
            return this;
        }

        public SubjectFieldMapperCatalog build() {
            return new SubjectFieldMapperCatalog(mappers);
        }
    }
}
